internal/collect/datacenters.go

```diff
@@ -33,47 +35,55 @@ func Datacenters(config *load.Config) {
 	}()
 
 	var datacenters []mo.Datacenter
-	err = cv.Retrieve(ctx, []string{"Datacenter"}, []string{"name", "overallStatus"}, &datacenters)
+	err = cv.Retrieve(ctx, []string{DATACENTER}, []string{"name", "overallStatus"}, &datacenters)
 	if err != nil {
-		config.Logrus.WithError(err).Fatal("failed to retrieve Datacenter")
+		config.Logrus.WithError(err).Error("failed to retrieve Datacenters")
+		return err
 	}
 
-	if config.Args.EnableVsphereTags && config.IsVcenterAPIType {
-		err := collectTagsByID(config.TagsByID, config.TagsManager)
+	var objectTags tag.TagsByObject
+	if config.TagCollectionEnabled() {
+		objectTags, err = config.TagCollector.FetchTagsForObjects(datacenters)
 		if err != nil {
-			config.Logrus.WithError(err).Error("failed to collect tags by id")
+			config.Logrus.WithError(err).Warn("failed to retrieve tags for datacenters")
 		}
 	}
 
+	filterByTag := config.TagFilteringEnabled()
+
+	// cache store for events
 	cs, err := newCacheStore(config)
 	if err != nil {
 		config.Logrus.WithError(err).Warn("could not create cache for vsphere events. all events will be returned")
 	}
-	for i, d := range datacenters {
-		newDatacenter := load.NewDatacenter(&datacenters[i])
-		if config.IsVcenterAPIType && config.Args.EnableVsphereEvents {
-			c := cache.NewCache(d.Name, cs)
-			collectEvents(config, d, newDatacenter, c)
-		}
 
-		if config.Args.EnableVspherePerfMetrics {
-			newDatacenter.PerfCollector, err = performance.NewPerfCollector(config.VMWareClient, config.Logrus, config.Args.PerfMetricFile, config.Args.LogAvailableCounters, config.Args.PerfLevel, config.Args.BatchSizePerfEntities, config.Args.BatchSizePerfMetrics)
-			if err != nil {
-				config.Logrus.Fatal(err)
-			}
+	for _, d := range datacenters {
+		if filterByTag && len(objectTags) == 0 {
+			config.Logrus.WithField("datacenter", d.Name).
+				Debugf("ignoring datacenter since not tags were collected and we have filters configured")
+			continue
 		}
-		config.Datacenters = append(config.Datacenters, newDatacenter)
+		// if object has no tags attached or no tag matches any of the tag filters, object will be ignored
+		if filterByTag && !config.TagCollector.MatchObjectTags(objectTags[d.Reference()]) {
+			config.Logrus.WithField("datacenter", d.Name).
+				Debugf("ignoring datacenter since it does not match any configured tag")
+			continue
+		}
```

Comment:
i think we should remark this in de docs, if filters are used datacenter should also have tags otherwise the hole objects that contains will be no collected

Reply:
Yes, correct. I'll make a note to change the docs